### PR TITLE
[xds] Rds in lis wrapper

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds RDS to Listener Wrapper. This is for any Filter Chains which specify dynamic RDS Responses. This PR is scoped only to Persisting RDS Configuration/Dynamic logic, and only the persistence in Listener Wrapper. Future functionality involves actually instantiating filters using this configuration, and matching (routing) + running through HTTP Filters in xds(Unary|Stream)Interceptor.

Note, this is branched off https://github.com/grpc/grpc-go/pull/4610.

RELEASE NOTES: None

### CODE REVIEW

**Comment by easwars** [comment]
Suggestion/Request: Please squash your commits before sending it our for review. The 40 odd commits take up a lot of real estate in this page.

File: xds/internal/server/rds_handler.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	rdsUpdates map[string]xdsclient.RouteConfigUpdate
+	err        error
+}
+
+// rdsHandler handles any RDS queries that need to be started for a given server
+// side listeners Filter Chains (i.e. not inline).
+type rdsHandler struct {
+	parent *listenerWrapper
+
+	rdsMutex sync.Mutex
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:39]: Given there are no other mutexes in this struct, `mu` would be preferable.

File: xds/internal/server/rds_handler.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+		}
+	}
+}
+
+// handleRouteUpdate persists the route config for a given route name, and also
+// sends an update to the Listener Wrapper on an error received or if the rds
+// handler has a full collection of updates.
+func (rh *rdsHandler) handleRouteUpdate(update xdsclient.RouteConfigUpdate, err error) {
+	rh.rdsMutex.Lock()
+	defer rh.rdsMutex.Unlock()
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:90]: Can we grab the lock after handling `err != nil`?

File: xds/internal/server/rds_handler.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+	// Note: this doesn't need a check for if name in RouteConfigUpdate is wrong
+	// (i.e. not started a watch for). This will be validated in xdsclient.
+	if err != nil {
+		// For a rdsHandler update, the only update wrapped listener cares about
+		// is most recent one, so opportunistically drain the update before
+		// sending the new update.
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:96]: Prefer placing this comment at the place where `updateChannel` is declared. That would save you from repeating the comment.

File: xds/internal/server/rds_handler.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	for routeName := range routeNamesToWatch {
+		if _, ok := rh.rdsCancels[routeName]; !ok {
+			rh.rdsCancels[routeName] = rh.parent.xdsC.WatchRouteConfig(routeName, rh.handleRouteUpdate)
+		}
+	}
+
+	// Delete and cancel watches for any routes from persisted routeNamesToWatch
+	// that are no longer present.
+	for routeName := range rh.rdsCancels {
+		if _, stillRDS := routeNamesToWatch[routeName]; !stillRDS {
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:77]: `ok` would be a better name.

File: xds/internal/server/rds_handler.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import (
+	"sync"
+
+	"google.golang.org/grpc/xds/internal/xdsclient"
+)
+
+// rdsHandlerUpdate wraps the full RouteConfigUpdate that are dynamically queried for a given
+// server side listener.
+type rdsHandlerUpdate struct {
+	rdsUpdates map[string]xdsclient.RouteConfigUpdate
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:30]: s/rdsUpdates/updates

File: xds/internal/server/rds_handler.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	// the listener is ready to be updated.
+	if len(rh.rdsUpdates) == len(rh.rdsCancels) {
+		// For a rdsHandler update, the only update lis wrapper cares about is most recent one,
+		// so opportunistically drain the update before sending the new update.
+		select {
+		case <-rh.updateChannel:
+		default:
+		}
+		rh.updateChannel <- rdsHandlerUpdate{rdsUpdates: rh.rdsUpdates}
+	}
```
Comment (easwars) [inline xds/internal/server/rds_handler.go:116]: Would the following case be covered?

- `updateRouteNamesToWatch(A, B, C)`
- `WatchRouteConfig()` is called for A, B and C
- We get a response for `A` and `B`
- Still the condition `len(rh.rdsUpdates) == len(rh.rdsCancels)` is not true, so no update is propagated to the `listenerWrapper`
- Then `updateRouteNamesToWatch(A, B)` happens
- map entries for `C` are deleted
- Now, the above condition is true, but the update will not be propagated up. Am I right?

I would prefer having a single map:
```golang
type updateWithCancel struct {
  update *xdsclient.RouteConfigUpdate
  cancel func()
}

type rdsHandler struct {
  ...
  updates map[string]updateWithCancel
  ...
}

func (rh *rdsHandler) readyToUpdate() bool { // or a better name
  // Traverses the map and figures out if we are ready to send an update upstairs
}
```

Call `readyToUpdate` from both `updateRouteNamesToWatch` and `handleRouteUpdate` and send update upstairs if required.

File: xds/internal/server/listener_wrapper.go [hunk 7]
```diff
@@ -84,6 +84,11 @@ func (s ServingMode) String() string {
 // non-nil error if the server has transitioned into not-serving mode.
 type ServingModeCallback func(addr net.Addr, mode ServingMode, err error)
 
+// DrainCallback is the callback that users can register to get pinged whenever
```
Comment (easwars) [inline xds/internal/server/listener_wrapper.go:87]: Who is the user here?

**Review by easwars — CHANGES_REQUESTED** [review]

File: xds/internal/server/rds_handler.go (hunk 6 above)
Comment (zasweq) [reply xds/internal/server/rds_handler.go:116]: From your perspective, why would a map[string]{A, B} be cleaner than map[string]{A} and a map[string]{B}? I agree with your correctness issue. However, traversing a single map would add a 0(n) search for every update/update route name to watch vs len() and len(), which is constant.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments!

**Comment by zasweq** [comment]
Closing this. This PR is continued in #4655.

**Pull request closed without merging**